lib/auth/middleware.go

```diff
@@ -668,22 +672,54 @@ func (a *Middleware) WrapContextWithUser(ctx context.Context, conn utils.TLSConn
 		}
 	}
 
-	return a.WrapContextWithUserFromTLSConnState(ctx, conn.ConnectionState())
+	return a.WrapContextWithUserFromTLSConnState(ctx, conn.ConnectionState(), conn.RemoteAddr())
 }
 
 // WrapContextWithUserFromTLSConnState enriches the provided context with the identity information
 // extracted from the provided TLS connection state.
-func (a *Middleware) WrapContextWithUserFromTLSConnState(ctx context.Context, tlsState tls.ConnectionState) (context.Context, error) {
+func (a *Middleware) WrapContextWithUserFromTLSConnState(ctx context.Context, tlsState tls.ConnectionState, remoteAddr net.Addr) (context.Context, error) {
 	user, err := a.GetUser(tlsState)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	ctx = context.WithValue(ctx, contextUserCertificate, certFromConnState(&tlsState))
+	ctx = context.WithValue(ctx, ContextClientAddr, remoteAddr)
 	ctx = context.WithValue(ctx, ContextUser, user)
 	return ctx, nil
 }
 
+// CheckIPPinning verifies IP pinning for the identity, using client ip take from context.
+// Check is considered successful if no error is returned.
+func CheckIPPinning(ctx context.Context, identity tlsca.Identity, pinSourceIP bool) error {
+	if identity.PinnedIP == "" {
```

Comment:
Should we check if `pinSourceIP` is false at the beginning and return? Or does the rest of this function still apply even if pinning is disabled?

Reply:
Yes, `PinnedIP` on certificate takes precedence. There are situation when user isn't required to be pinned by roles but certificate should still be pinned - in case of single use certificates.